src/string_id.h

```diff
@@ -69,69 +107,96 @@ template<typename T>
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
+            _is_char_ptr( other._is_char_ptr ) {}
+
         template<typename S, class = typename
-                 std::enable_if< std::is_convertible<S, std::string >::value>::type >
-        explicit string_id( S &&
-                            id ) : _id( std::forward<S>( id ) ), _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+                 std::enable_if< std::is_convertible< S, std::string >::value >::type
+                 >
+        explicit string_id( S && id ) {
+            auto str = std::make_unique<std::string>( std::forward<S>( id ) );
+            uint64_t hash = Hash::RunTime::fnv1_64( str->c_str(), str->size() );
+            _hash = hash;
+
+            auto insertion = cached_strings().try_emplace( hash, std::move( str ) );
+            auto cached_string = insertion.first->second.get();
+
+            // Verify we don't have hash collisions
+            //assert(insertion.second || *cached_string == id);
+
+            _string = cached_string;
+        }
+
         /**
          * Default constructor constructs an empty id string.
          * Note that this id class does not enforce empty id strings (or any specific string at all)
          * to be special. Every string (including the empty one) may be a valid id.
          */
-        string_id() : _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+        constexpr string_id() : string_id( "" ) {};
         /**
          * Comparison, only useful when the id is used in std::map or std::set as key. Compares
          * the string id as with the strings comparison.
          */
         bool operator<( const This &rhs ) const {
-            return _id < rhs._id;
+            // Not str() compares to prevent needless tier ups to std::string
+            return strcmp( c_str(), rhs.c_str() ) < 0;
         }
         /**
          * The usual comparator, compares the string id as usual.
          */
         bool operator==( const This &rhs ) const {
-            bool can_compare_cid = ( _cid != INVALID_CID || rhs._cid != INVALID_CID ) &&
-                                   _version == rhs._version && _version != INVALID_VERSION;
-            return ( can_compare_cid && _cid == rhs._cid ) ||
-                   ( !can_compare_cid && _id == rhs._id );
-            // else returns false, when:
-            //      can_compare_cid && _cid != rhs._cid
-            //  OR  !can_compare_cid && _id != rhs._id
+            return _hash == rhs._hash;
```

Comment:
String comparison using hashes is not guaranteed to be correct. Even if it's possible to prove that all currently used ids have different hashes, it still doesn't guarantee that there won't be a scenario in the future when dynamically generated ids produce a hash collision, leading to a very obscure bug.

Reply:
While it isn't a solution, a check can be put in place.
Whenever we create a string_id, we try to insert it and get a pointer to the existing string if its hash already appeared before.
If I were to add a check which would at the very least throw a warning when such a case happens, would you consider that a solution? (Something slower can be put in place, such as inc-ing the hash till an unused one is reached, but it is pretty ugly)

Reply:
Frankly, I don't know. While I realize that the chance of collision is slim, what should we do if one happens, even if we can successfully detect it in runtime? There are lots of mods on the internet, each can use arbitrary ids. If a player using a specific set of mods steps on a collision, as a dev team that has to support this, what would be our next steps?

Also, I don't think that currently the check works, as you will potentially have multiple `cached_strings_` instances even for a single id type.

Reply:
I think you can fall back to string comparison if the hashes are the same, which should avoid the hash collision.

Reply:
> I think you can fall back to string comparison if the hashes are the same, which should avoid the hash collision.

Then the very common case when string_id is looked up in the collection that contains it will be still slow, i.e. look up in the `map` or `unordered_map` will result in string comparison, which we're trying to avoid.

Reply:
String comparison will only happen when the element is found (assuming no hash collision happens), so it's still faster than before. Although now I think Aviean's suggestion of replacing the hash with the unique pointer might be a better idea.